### ISSUE TITLE
Circular dependency avoidance

### ISSUE THREAD

**Issue by bcardarella** [elixir-lang/elixir#4420]
I'm not sure if this would be considered a bug or not. Here is my situation.

I have a library I wrote. It is adapter based. One of the adapters was getting to be large so I broke it out into its own library. Let's call the original library the Parent and the extracted library the Child.

In the child library I depend upon the parent library in `mix.exs`. However, I want to include the child library along with the parent library, I just want to keep the repos separate. (the child library was accounting for 99% of the test suite time and will see much less churn/releases than the parent)

So I set the child as a dependency of the parent.

Child `mix.exs`:

```elixir
{:parent, "~> 0.01"}
```

Parent `mix.exs`:

```elixir
{:child, "~> 0.0.1"}
```

I figured that Elixir would just resolve the circular dependency. The problem is that when I run`mix deps.get` on the parent it pulls down the published child library, puts it into `deps/child`. Then when I try to run `mix test` on the parent I get the following:

```
Error while loading project :courier at /Users/bcardarella/Dropbox/projects/courier/deps/courier
** (Mix) Trying to load Courier.Mixfile from "/Users/me/projects/parent/deps/parent/mix.exs" but another project with the same name was already defined at "/Users/me/projects/parent/mix.exs"
```

Basically the copy of the parent in `deps/` is conflicting with the development copy that I attempting to test against.

Thoughts?

**Comment by josevalim** [comment]
You can't have circular dependencies. Maybe we can raise better error messages but we certainly can't make it work.

**Comment by wojtekmach** [comment]
@bcardarella I've became a fan of monorepos of late, so please take my advice with a grain of salt :) I think keeping "officially supported" adapters within the same repo/release is usually a better experience for both maintainers and end-users. I can't speak for the maintainers of these projects, but there are popular examples that follow "batteries included" pattern like e.g.: Ecto and Terraform. For me the biggest problem with maintaining 1+ repo for similar stuff was "synchronizing" PRs.

As far as slow test suite for a particular adapter, I think there's always an option to skip it by default and always run it on CI. I've also worked in the past on scripts for CI PR builds that check which files were changed on the branch and only run tests for those.

Perhaps an umbrella app would be an option too? I haven't used umbrella apps for projects that are supposed to be dependencies of other projects, but I guess it should work - and you'd get a nice way of running tests per app out of the box. There's also an option to extract the essential parts of parent that the child needs into a separare dep (e.g. parent_adapter) that both child and parent (through child) depend on. Shameless plug: you can see how this could work in practice here: https://github.com/wojtekmach/xeh

Hope it was helpful.

**Comment by bcardarella** [comment]
@josevalim in this case it seems odd that the child having the parent as the dep would still install the parent into `deps/` if being called from the parent itself. But I hear you, I figured it was a long shot

**Comment by bcardarella** [comment]
@wojtekmach I've debated it, I think it will add a lot of headache. I suspect in the case I will just make the child dep an opt-in
